src/ipred.rs

```diff
@@ -1024,50 +1023,47 @@ unsafe fn ipred_z2_rust<BD: BitDepth>(
                 height,
                 topleft_in,
                 topleft_in_off - height as usize,
-                0 as c_int,
+                0,
                 height + 1,
                 filter_strength_0,
             );
         } else {
+            let height = height as usize;
             BD::pixel_copy(
-                &mut edge[topleft - height as usize..][..height as usize],
-                &topleft_in[topleft_in_off - height as usize..][..height as usize],
-                height.try_into().unwrap(),
+                &mut edge[topleft - height..][..height],
+                &topleft_in[topleft_in_off - height..][..height],
+                height,
             );
         }
     }
     edge[topleft] = topleft_in[topleft_in_off];
-    let base_inc_x = 1 + upsample_above as c_int;
-    let left = edge[topleft - (1 + upsample_left as usize)..].as_mut_ptr();
-    let mut y = 0;
-    let mut xpos = (1 + (upsample_above as c_int) << 6) - dx;
-    while y < height {
-        let mut base_x = xpos >> 6;
-        let frac_x = xpos & 0x3e as c_int;
-        let mut x = 0;
-        let mut ypos = (y << 6 + upsample_left as c_int) - dy;
-        while x < width {
-            let v;
-            if base_x >= 0 {
-                v = edge[topleft + base_x as usize].as_::<c_int>() * (64 - frac_x)
-                    + edge[topleft + base_x as usize + 1].as_::<c_int>() * frac_x;
+
+    let base_inc_x = 1 + upsample_above as usize;
+    let left = topleft - (1 + upsample_left as usize);
+    for y in 0..height {
+        let xpos = (1 + (upsample_above as c_int) << 6) - (dx * (y + 1));
+        let base_x = xpos >> 6;
+        let frac_x = xpos & 0x3e;
+
+        let dst_slice = slice::from_raw_parts_mut(
+            dst.offset(BD::pxstride(stride) * y as isize),
+            width as usize,
+        );
+        for (x, dst) in dst_slice.iter_mut().enumerate() {
+            let ypos = (y << 6 + upsample_left as c_int) - (dy * (x + 1) as c_int);
+            let base_x = base_x + (base_inc_x * x) as c_int;
+            let v = if base_x >= 0 {
+                edge[topleft + base_x as usize].as_::<c_int>() * (64 - frac_x)
+                    + edge[topleft + base_x as usize + 1].as_::<c_int>() * frac_x
```

Comment:
Can you pre-slice `&edge[topleft..]`?

Reply:
I don't think that makes this any cleaner, I think I prefer the consistency of using `base + offset` since we have to do the same thing in the other branch with `left`.

Reply:
It is a bit more efficient this way, though.